core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -784,7 +784,7 @@ object Doc {
               val first = Concat(flatx, Concat(flatSep, resty))
               val second = Concat(x, Concat(sep, defer(cheatRec(y, tail))))
               // note that first != second
-              Union(first, second)
+              Union(first.flatten, second)
```

Comment:
I think this could really harm performance. Note that `first = Concat(flatx, Concat(flatSep, resty))` so, the first part is already flattened.

Why do we need that the left side is fully flattened? I don't see that requirement.

The requirement I think we need is for correctness is that the left side has a longer first line than the right so that we should check it first and if it can't fit, only then go to the right.

I think this construction already follows that rule. I don't want to give up the performance of this method unless we really need to.

Reply:
Oh, ha, I assumed this invariant from the comment:

```
   * Additionally, the left side (a) MUST be right associated with
   * any Concat nodes to maintain efficiency in rendering. This
   * is currently done by flatten/flattenOption.
```

added in `Oscar Boykin         2017-04-08 12:06 Remove right-association in def +`

I am happy to drop the requirement.  I thought it may be required due to the non-lazy-by-default nature of Scala vs. Haskell (which only maintains the length requirement you mention).  Can we make that comment less definitive-sounding?

Reply:
Isn’t the left side right associated? It is not done by fully flattening but I think it is right associated

Reply:
Sure, I'm being very sloppy, but look at the error message.  `Message: Left-associative Concat`

Reply:
Can you use representation to print it?

I guess it can be from sep not being right associated and we don’t reassociate that.

Looking at the algorithm it seems we should be able to associate it correctly fairly minimally in terms of work.

Reply:
```
- Union invariants *** FAILED *** (121 milliseconds)
[info]   TestFailedException was thrown during property evaluation.
[info]     Message: Left-associative Concat: Concat(Concat(Text( ), Empty), Concat(Text( ), Empty))
[info]     Location: (PaigesScalacheckTest.scala:331)
[info]     Occurred when passed generated values (
[info]       arg0 = Empty, // 10 shrinks
[info]       arg1 = Concat(Union(Text( ), Line(true)), Empty), // 1 shrink
[info]       arg2 = Empty
[info]     )
```

`fill` also generates lots of `Union`s on the left of `Union`s.  This seems like it should be called out in the comments/invariants as well.

Reply:
So, I think the issue here is that the last doc is not touched at all in the fill loop, and that is fed through the call stack. That may or may not be a bug. I think maybe the last thing sent though should be the grouped version of the the last thing in the list, then I *think* we will satisfy the invariant.

We probably need more tests on fill to be clearer about what it is doing. You made changes before and no tests failed.

if you look at the bottom of page 14: http://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf you see the definition:
```haskell
fill []        =  nil
fill [x]       =  x
fill (x:y:zs)  =  (flatten x <+> fill (flatten y : zs)) :<|> (x </> fill (y : zs))
```
but this is a weird definition that explicitly handles the last doc separately.

If you remove the middle clause `fill [x] = x` you would instead have `fill x : nil = flatten x :<|> x = grouped x` which seems more natural to me, and would, I think pass the invariant we want.

There is no explanation given for this exception to the rule.

Do you want to make that change on line 762
```scala
case Nil => call(x.grouped, stack)
```

If we make that change, and maybe try to express laws for fill, that would be pretty nice.

Reply:
Nice analysis!  Yeah, that change worked.   I was surprised a bit by having `Union`s on the lhs of other `Union`s, but it's clear in the paper that that would naturally happen there in the Haskell implementation as well.  I'll combine #115 and #110 in a new review, and drop those.

Reply:
Great news!